lib/Optimizer/Transforms/LiftArrayAlloc.cpp

```diff
@@ -27,6 +27,64 @@ namespace cudaq::opt {
 
 using namespace mlir;
 
+namespace {
+template <typename A>
+std::vector<A> readConstantValues(SmallVectorImpl<Attribute> &vec, Type eleTy) {
+  std::vector<A> result;
+  for (auto a : vec) {
+    if constexpr (std::is_same_v<A, std::complex<double>>) {
+      auto v = cast<ArrayAttr>(a);
+      result.emplace_back(cast<FloatAttr>(v[0]).getValue().convertToDouble(),
+                          cast<FloatAttr>(v[1]).getValue().convertToDouble());
+    } else if constexpr (std::is_same_v<A, std::complex<float>>) {
+      auto v = cast<ArrayAttr>(a);
+      result.emplace_back(cast<FloatAttr>(v[0]).getValue().convertToFloat(),
+                          cast<FloatAttr>(v[1]).getValue().convertToFloat());
+    } else if constexpr (std::is_same_v<A, double>) {
+      auto v = cast<FloatAttr>(a);
+      result.emplace_back(v.getValue().convertToDouble());
+    } else if constexpr (std::is_same_v<A, float>) {
+      auto v = cast<FloatAttr>(a);
+      result.emplace_back(v.getValue().convertToFloat());
+    } else {
+      assert(false && "unexpected type in constant array");
+    }
+  }
+  return result;
+}
+
+void genVectorOfConstantsFromAttributes(cudaq::IRBuilder irBuilder,
+                                        Location loc, ModuleOp module,
+                                        StringRef name,
+                                        SmallVector<Attribute> &values,
+                                        Type eleTy) {
+
+  if (auto cTy = dyn_cast<ComplexType>(eleTy)) {
+    auto floatTy = cTy.getElementType();
+    if (floatTy == irBuilder.getF64Type()) {
+      auto vals = readConstantValues<std::complex<double>>(values, cTy);
+      irBuilder.genVectorOfConstants(loc, module, name, vals);
+      return;
+    } else if (floatTy == irBuilder.getF32Type()) {
+      auto vals = readConstantValues<std::complex<float>>(values, cTy);
+      irBuilder.genVectorOfConstants(loc, module, name, vals);
+      return;
+    }
+  } else if (auto floatTy = dyn_cast<FloatType>(eleTy)) {
+    if (floatTy == irBuilder.getF64Type()) {
+      auto vals = readConstantValues<double>(values, floatTy);
+      irBuilder.genVectorOfConstants(loc, module, name, vals);
+      return;
+    } else if (floatTy == irBuilder.getF32Type()) {
+      auto vals = readConstantValues<float>(values, floatTy);
+      irBuilder.genVectorOfConstants(loc, module, name, vals);
+      return;
+    }
+  }
+  assert(false && "unexpected element type in constant array");
```

Comment:
I don't think this pass should crash if it sees more general code that it is unsafe to transform in this way.

Reply:
Thanks, will update it to continue

Reply:
Fixed! Now the pass will just create a const array instead of failing to create a global.